data-raw/splenocytes.R

```diff
@@ -1,21 +1,108 @@
-
 library(djvdj)
 library(Seurat)
 library(here)
 library(tidyverse)
 library(Matrix)
 library(Rsamtools)
+library(clustifyr)
+library(clustifyrdata)   # USE EXPERIMENTHUB INSTEAD
+
+
+# Functions ----
+
+# Demultiplex samples
+demux_so <- function(dat_dir) {
+
+  # Create Seurat object
+  dat_dir <- file.path(
+    dat_dir,
+    "JH179_GEX-JH181_ADT_JH181_HTO/outs/filtered_feature_bc_matrix"
+  )
+
+  mat <- Read10X(dat_dir)
 
+  so <- mat$`Gene Expression` %>%
+    CreateSeuratObject()
 
+  # Add HTO data and demultiplex
+  hto_mat <- mat$`Antibody Capture`
+  hto_mat <- hto_mat[rownames(hto_mat) != "_HEL", ]
 
-# ---- WRITE RAW MATRICES
+  so[["HTO"]] <- CreateAssayObject(hto_mat)
 
-# Create Seurat object and demultiplet samples
-# Save a separate matrix for each sample
+  so <- so %>%
+    NormalizeData(assay = "HTO", normalization.method = "CLR") %>%
+    HTODemux(assay = "HTO") %>%
+    subset(hash.ID != "Doublet" & hash.ID != "Negative")
 
-dat_dir <- c("~/Projects/Smith_AVIDseq/results/2020-01-18/")
+  # Preprocess object
+  so <- so %>%
+    preprocess_so()
+
+  so
+}
+
+# Preprocess object
+preprocess_so <- function(so_in) {
+  res <- so_in %>%
+    NormalizeData() %>%
+    FindVariableFeatures(nfeatures = 1000) %>%
+    ScaleData() %>%
+    RunPCA() %>%
+    RunUMAP(dims = 1:30) %>%
+    FindNeighbors(dims = 1:30) %>%
+    FindClusters(resolution = 1)
+
+  # Annotate B/T cells
+  res <- res %>%
+    clustify(
+      cluster_col = "seurat_clusters",
+      ref_mat = ref_immgen
+    )
+
+  res <- res %>%
+    mutate_meta(
+      mutate,
+      cell_type = str_remove(type, " \\(.+$"),
+      cell_type = recode(cell_type, Tgd = "T cells"),
+      cell_type = ifelse(
+        cell_type %in% c("B cells", "T cells"),
+        cell_type,
+        "other"
+      )
+    )
+
+  res
+}
+
+# Downsample object
+downsample_so <- function(so_in, n, b_frac = 0.8) {
+
+  if (is.null(n) || is.null(b_frac)) return(so_in)
+
+  # Sample B and T cells
+  n_b <- n * b_frac
+  n_t <- n - n_b
+
+  b_bcs <- so %>%
```

Comment:
would be good to set a seed before each `sample()` to make data generation reproducible